### PR TITLE
Add implementation details to CCM migration KEP

### PR DESCRIPTION
**Pull request by andrewsykim**
Signed-off-by: andrewsykim <kim.andrewsy@gmail.com>

### CODE REVIEW

File: keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md [hunk 1]
```diff
@@ -152,53 +157,125 @@ will prevent any of the v1.18 CCMs from claiming the lock. When the current hold
...
-### Graduation Criteria
+	// Component is the name of the component in which the controller should be running.
+	// E.g. kube-controller-manager, cloud-controller-manager, etc
+	Component string `json:"component"`
+}
+```
+
+#### Component Flags
+
+The LeaderMigrationConfiguration type will be read by the `kube-controller-manager` and the `cloud-controller-manager` via a new flag `--cloud-migration-config` which
```
Comment (andrewsykim) [inline keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md]: Maybe just `--controller-migration-config` to keep it generic?

File: keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md [hunk 2]
```diff
@@ -152,53 +157,125 @@ will prevent any of the v1.18 CCMs from claiming the lock. When the current hold
...
+#### Migration Configuration
 
-### Risks and Mitigations
+The migration lock will be configured by defining new API types that will then be passed into the KCM and CCM.
 
-TODO after KEP summary and proposal is approved
+```go
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// LeaderMigrationConfiguration provides versioned configuration for all migrating leader locks.
+type LeaderMigrationConfiguration struct {
```
Comment (andrewsykim) [inline keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md]: I don't feel strongly about how this API is designed, I'll defer to @cheftako on this

File: keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md [hunk 3]
```diff
@@ -152,53 +160,125 @@ will prevent any of the v1.18 CCMs from claiming the lock. When the current hold
...
+
+First, define a LeaderMigrationConfiguration resource in a yaml file containing all known cloud controllers. The component name for each controller should be set to
+the component where the controllers are currently running. Almost always this is the `kube-controller-manager`. The configuration file should look something like this:
+```yaml
+kind: LeaderMigrationConfiguration
+apiVersion: v1alpha1
+leaderName: cloud-controllers-migration
+controllerLeaders:
+  - name: route-controller
+    component: kube-controller-manager
```
[the remark this answers is not in the document]
Comment (andrewsykim) [reply keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md]: Not sure I agree that `leader` makes more sense for this field. I like `component` cause it's more clearer it should be either  `kube-controller-manager` or `cloud-controller-manager`. `leader` is a bit more ambiguous

File: keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md [hunk 4]
```diff
@@ -152,53 +160,125 @@ will prevent any of the v1.18 CCMs from claiming the lock. When the current hold
...
+
+##### Enable Leader Migration on Components
+
+First, define a LeaderMigrationConfiguration resource in a yaml file containing all known cloud controllers. The component name for each controller should be set to
+the component where the controllers are currently running. Almost always this is the `kube-controller-manager`. The configuration file should look something like this:
+```yaml
+kind: LeaderMigrationConfiguration
+apiVersion: v1alpha1
+leaderName: cloud-controllers-migration
+controllerLeaders:
```
[the remark this answers is not in the document]
Comment (andrewsykim) [reply keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md]: I'll defer to @cheftako on this one

File: keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (yastij) [reply keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md]: @andrewsykim - I guess this scoped for the ccm migration and versioned API,  so we can revisit this later If we want to expand this for general purpose

[the remark this answers is not in the document]
Comment (yastij) [reply keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md]: @andrewsykim - I guess this scoped for the ccm migration and versioned API,  so we can revisit this later If we want to expand this for general purpose

File: keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md (hunk 1 above)
Comment (cheftako) [reply keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md]: +1

File: keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (cheftako) [reply keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md]: I agree. I find leader here a bit confusing.

File: keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (cheftako) [reply keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md]: +1. I'd like to see the change but this is also alpha.

File: keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (cheftako) [reply keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md]: Or possible controllerToLeader?

File: keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md [hunk 5]
```diff
@@ -152,53 +160,125 @@ will prevent any of the v1.18 CCMs from claiming the lock. When the current hold
...
+// ControllerLeaderConfiguration provides the configuration for a migrating leader lock.
+type ControllerLeaderConfiguration struct {
+	// Name is the name of the controller being migrated
+	// E.g. service-controller, route-controller, cloud-node-controller, etc
+	Name string `json:"name"`
 
-### Graduation Criteria
+	// Component is the name of the component in which the controller should be running.
+	// E.g. kube-controller-manager, cloud-controller-manager, etc
+	Component string `json:"component"`
```
Comment (cheftako) [inline keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md]: If we are calling this leader elsewhere, then we should probably call it leader here.

**Pull request merged**